torch/futures/__init__.py

```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
     Wrapper around a ``torch._C.Future`` which encapsulates an asynchronous
     execution of a callable, e.g. :meth:`~torch.distributed.rpc.rpc_async`. It
     also exposes a set of APIs to add callback functions and set results.
+
+    .. warning:: GPU support is a beta feature, subject to changes.
     """
 
     def done(self) -> bool:
         r"""
         Return ``True`` if this ``Future`` is done. A ``Future`` is done if it
         has a result or an exception.
+
+        If the value contains tensors that reside on GPUs, ``Future.done()``
+        will return ``True`` even if the asynchronous kernels that are
+        populating those tensors haven't yet completed running on the device,
+        because at such stage the result is already usable, provided one
+        performs the appropriate synchronizations (see :meth:`wait`).
         """
         return super().done()
 
     def wait(self, non_blocking: bool = False) -> T:
         r"""
         Block until the value of this ``Future`` is ready.
 
+        If the value contains tensors that reside on GPUs, then an additional
+        synchronization is performed with the kernels (executing on the device)
+        which may be asynchronously populating those tensors. Such sync can be:
+
+        - *Blocking* (the default), in which case the calling thread is held up
+          until those asynchronous kernels effectively terminate running on the
+          device. In such a mode, when ``wait()`` returns it is safe to
+          immediately use the result on any stream without further
+          synchronization. On the other hand, a blocking wait delays the moment
+          at which new operations can be sent to the GPU, possibly degrading
+          performance.
+        - *Non-blocking*, in which case ``wait()`` inserts the necessary
+          instructions in the current streams to ensure that further operations
+          enqueued on those streams will be properly scheduled after the async
+          kernels. Once that is done, ``wait()`` will return, even if
+          those kernels are still running. This mode allows for better
+          pipelining and performance, and should be safe as long as one doesn't
+          change streams between the call to ``wait()`` and the time at
+          which the result is used (but it requires explicit manual
+          synchronization when switching streams).
+
+        Args:
+            non_blocking (bool): whether the sync with the GPU kernels should be
+                                 non-blocking.
+
         Returns:
             The value held by this ``Future``. If the function (callback or RPC)
             creating the value has thrown an error, this ``wait`` method will
             also throw an error.
         """
         return super().wait(non_blocking)
 
+    def value(self) -> T:
```

Comment:
I'm not adding this method, it already existed (implicitly inherited from the superclass) but it was undocumented.

Reply:
This LGTM, but I wonder if we should call this `result()` to 1) stay consistent with Python result 2) we already have a `set_result` API.

Reply:
Yep, just a general question on whether it worths to retire `value` and rename it `result`